presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java

```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
 
     public void updateMemoryUsage(long bytesAdded)
     {
-        SettableFuture<?> future;
-        synchronized (this) {
-            bufferedBytes.addAndGet(bytesAdded);
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                //  if we have no callback waiting, return early
+                if (notFullFuture == NOT_FULL) {
```

Comment:
use `notFullFuture.isDone()`

Reply:
`NOT_FULL` is the sentinel value for "no listeners". Any other value is a future that might not be done, but that check can occur outside of the synchronized block and without stalling to do a volatile read against that future's `isDone()` state. Redudant `set(null)` calls are benign but this implementation preserves the ability to alias check the `NOT_FULL` sentinel to detect listener registration.

Reply:
Replaced with null check instead.